lib/Core/Executor.cpp

```diff
@@ -905,14 +905,21 @@ Executor::fork(ExecutionState &current, ref<Expr> condition, bool isInternal) {
       }
     }
 
+    // We don't add a constraint explicitly yet
```

Comment:
@ccadar  @MartinNowack @kferles I see. Right now it is up to the strategy to do the "right" thing. Looking at the possibilities, it seems like for DFS and BFS it should be (2), (3), or (4) since depth is usually not measured in terms of instructions. As I said earlier, (3) kind of feels like the most intuitive choice to me.

For random, it seems like it depends on what you want to pick randomly. Based on Cristian's description I would call the current strategy "random instruction", but I could definitely imagine something like a "random branch" strategy as well. In fact, that's what I wrongly assumed the random strategy was doing.

The liveness/fairness issue is somewhat orthogonal. One could still support something like the safety net I suggested as a simple wrapper for any strategy.

I think that the most important thing is to make sure this is made more explicit in the documentation and the code. For instance, it seems like it might help to pass an argument to the update method that indicates where (branch, symbolic branch, 2-way fork, other instruction) the update occurred. A strategy like BFS, could then make such decisions without tracking added constraints or making implicit assumptions about some invariants that should hold about the arguments.